### PR TITLE
Force `mode2` to always be string

### LINKED ISSUE #4094: Make sure Mode2 can only be a string
Currently, the type allows for strings or numbers, which can be confusing when comparing. I think it would be better to force it to always be a string

### PR DESCRIPTION
**Pull request by Ferotiq**
### Description

Closes #4094

### CODE REVIEW

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/4480782636

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/4495217618

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/4495351943

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/4495415637

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/4495512859

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/4495591296

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/4495730795

File: backend/src/types/types.d.ts [hunk 1]
```diff
@@ -350,17 +352,11 @@ declare namespace MonkeyTypes {
...
-      [key: number]: PersonalBest[];
-    };
-    quote: { [quote: string]: PersonalBest[] };
-    custom: { custom?: PersonalBest[] };
-    zen: {
-      zen?: PersonalBest[];
-    };
+    time: Partial<Record<StringNumber, PersonalBest[]>>;
+    words: Partial<Record<StringNumber, PersonalBest[]>>;
+    quote: Partial<Record<string, PersonalBest[]>>;
```
Comment (Miodec) [inline backend/src/types/types.d.ts:357]: Why are you not using `StringNumber` here?

File: backend/src/types/types.d.ts [hunk 2]
```diff
@@ -350,17 +352,11 @@ declare namespace MonkeyTypes {
...
+    custom: Partial<Record<"custom", PersonalBest[]>>;
+    zen: Partial<Record<"zen", PersonalBest[]>>;
   }
```
Comment (Miodec) [inline backend/src/types/types.d.ts:360]: Also, why is `Partial` needed here?

File: frontend/src/ts/account/pb-tables.ts [hunk 3]
```diff
@@ -93,8 +93,10 @@ export function update(
   $(`.page${source} .profile .pbsTime`).html("");
   $(`.page${source} .profile .pbsWords`).html("");
 
-  text = "";
```
Comment (Miodec) [inline frontend/src/ts/account/pb-tables.ts:96]: Is this line not needed?

File: frontend/src/ts/test/test-logic.ts [hunk 4]
```diff
@@ -1579,6 +1579,8 @@ export async function finish(difficultyFailed = false): Promise<void> {
   let afkDetected = kps.every((second) => second.afk);
   if (TestInput.bailout) afkDetected = false;
 
+  const mode2Value = parseInt(completedEvent.mode2);
```
Comment (Miodec) [inline frontend/src/ts/test/test-logic.ts:1582]: Rename to `mode2Number` ?

File: frontend/src/ts/types/types.d.ts [hunk 5]
```diff
@@ -292,17 +292,11 @@ declare namespace MonkeyTypes {
...
-    quote: { [quote: string]: PersonalBest[] };
-    custom: { custom: PersonalBest[] };
-    zen: {
-      zen: PersonalBest[];
-    };
+    time: Partial<Record<StringNumber, PersonalBest[]>>;
+    words: Partial<Record<StringNumber, PersonalBest[]>>;
+    quote: Partial<Record<string, PersonalBest[]>>;
+    custom: Partial<Record<"custom", PersonalBest[]>>;
+    zen: Partial<Record<"zen", PersonalBest[]>>;
```
Comment (Miodec) [inline frontend/src/ts/types/types.d.ts:299]: Same comments as the backend types

File: frontend/src/ts/account/pb-tables.ts (hunk 3 above)
Comment (Ferotiq) [reply frontend/src/ts/account/pb-tables.ts:96]: No, it's initialized to `""` already.

File: backend/src/types/types.d.ts [hunk 6]
```diff
@@ -350,17 +352,11 @@ declare namespace MonkeyTypes {
...
-    quote: { [quote: string]: PersonalBest[] };
-    custom: { custom?: PersonalBest[] };
-    zen: {
-      zen?: PersonalBest[];
-    };
+    time: Record<StringNumber, PersonalBest[]>;
+    words: Record<StringNumber, PersonalBest[]>;
+    quote: Record<StringNumber, PersonalBest[]>;
+    custom: Partial<Record<"custom", PersonalBest[]>>;
+    zen: Partial<Record<"zen", PersonalBest[]>>;
```
Comment (Miodec) [inline backend/src/types/types.d.ts:359]: Why is `Partial` needed for custom and zen?

File: frontend/src/ts/elements/leaderboards.ts [hunk 7]
```diff
@@ -16,56 +16,56 @@ let currentTimeRange: "allTime" | "daily" = "allTime";
...
 
 let updateTimer: number | undefined;
 
-function clearBody(lb: number): void {
-  if (lb === 15) {
+function clearBody(lb: MonkeyTypes.StringNumber): void {
+  if (lb === "15") {
     $("#leaderboardsWrapper table.left tbody").empty();
-  } else if (lb === 60) {
+  } else if (lb === "60") {
```
[the remark this answers is not in the document]
Comment (Miodec) [reply frontend/src/ts/elements/leaderboards.ts:47]: Or just set it to `LbKey`

File: backend/src/types/types.d.ts (hunk 6 above)
Comment (Ferotiq) [reply backend/src/types/types.d.ts:359]: Aren't the custom and zen fields optional?

File: frontend/src/ts/elements/leaderboards.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ferotiq) [reply frontend/src/ts/elements/leaderboards.ts:47]: mk

File: backend/src/types/types.d.ts (hunk 6 above)
Comment (Miodec) [reply backend/src/types/types.d.ts:359]: Techincally, all the PersonalBests are optional

Comment (Miodec) [reply backend/src/types/types.d.ts:359]: I think anway

**Review by Miodec — APPROVED** [review]

**Pull request merged**